sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs

```diff
@@ -683,6 +686,83 @@ public virtual async Task CloseAsync(CancellationToken cancellationToken = defau
+        /// The dead letter reason and error description can only be specified either through the method parameters or hard coded
+        /// using this properties.
+        /// </remarks>
+        /// <exception cref="ServiceBusException">
+        ///   <list type="bullet">
+        ///     <item>
+        ///       <description>
+        ///         The lock for the message has expired or the message has already been completed. This does not apply for session-enabled entities.
+        ///         The <see cref="ServiceBusException.Reason" /> will be set to <see cref="ServiceBusFailureReason.MessageLockLost"/> in this case.
+        ///       </description>
+        ///     </item>
+        ///     <item>
+        ///       <description>
+        ///         The lock for the session has expired or the message has already been completed. This only applies for session-enabled entities.
+        ///         The <see cref="ServiceBusException.Reason" /> will be set to <see cref="ServiceBusFailureReason.SessionLockLost"/> in this case.
+        ///       </description>
+        ///     </item>
+        ///   </list>
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        ///   <list type="bullet">
+        ///     <item>
+        ///       <description>
+        ///         The dead letter reason or dead letter error exception was specified in both the parameter and the properties dictionary.
+        ///       </description>
+        ///     </item>
+        ///   </list>
+        /// </exception>
+        public virtual async Task DeadLetterMessageAsync(
+            ServiceBusReceivedMessage message,
+            IDictionary<string, object> propertiesToModify,
+            string deadLetterReason,
+            string deadLetterErrorDescription = default,
+            CancellationToken cancellationToken = default)
+        {
+            Argument.AssertNotNull(message, nameof(message));
+            Argument.AssertNotNull(propertiesToModify, nameof(propertiesToModify));
+
+            // Prevent properties and arguments from setting distinct deadletter reasons or error descriptions
+            var containsReasonHeader = propertiesToModify.TryGetValue(AmqpMessageConstants.DeadLetterReasonHeader, out var reasonHeaderProperty);
+            var containsDescriptionHeader = propertiesToModify.TryGetValue(AmqpMessageConstants.DeadLetterErrorDescriptionHeader, out var descriptionHeaderProperty);
+
+            if ((containsReasonHeader && deadLetterReason != null) && (reasonHeaderProperty.ToString() != deadLetterReason))
+            {
+                throw new InvalidOperationException("Differing deadletter reasons cannot be passed through both the properties and the parameter. Pass this value in one or the other place.");
+            }
+
+            if ((containsDescriptionHeader && deadLetterErrorDescription != null) && (descriptionHeaderProperty.ToString() != deadLetterErrorDescription))
```

Comment:
```suggestion
            if ((containsDescriptionHeader && deadLetterErrorDescription != null) && (descriptionHeaderProperty != deadLetterErrorDescription))
```

Reply:
Calling ToString would either be a no-op in the case that the variable was already a string, or it would end up calling ToString on some reference type like an array. I think by just comparing the instances directly we get the behavior that we want - if they are both strings we get a value comparison, if one is not a string then we return true.